### PR TITLE
Implements support for promise based migrations.

### PR DESCRIPTION
**Pull request by jdpnielsen**

### CODE REVIEW

File: lib/migrate.js [hunk 1]
```diff
@@ -42,6 +47,30 @@ function migrate (set, direction, migrationName, fn) {
...
+
+          // Decrement last run index
+          lastRunIndex--
+
+          set.lastRun = direction === 'up' ? migration.title : set.migrations[lastRunIndex] && set.migrations[lastRunIndex].title
+          set.save(function (err) {
+            if (err) return fn(err)
+
+            next(migrations.shift())
+          })
```
Comment (jdpnielsen) [inline lib/migrate.js]: Is this fine, or should i split the callback out into a separate function, which can be reused in both callback & promise cases?

Comment (wesleytodd) [reply lib/migrate.js]: Yeah, I would do that for sure.

File: lib/migrate.js [hunk 2]
```diff
@@ -42,6 +47,30 @@ function migrate (set, direction, migrationName, fn) {
         next(migrations.shift())
       })
     })
+
+    if (migrationPromise && migrationPromise.then) {
```
Comment (wesleytodd) [inline lib/migrate.js]: Prefer this to be `typeof Promise !== 'undefined' && migrationPromise instanceof Promise`

**Comment by wesleytodd** [comment]
Would prefer to keep as much backward support as possible.  Can you revert the `let`'s to `var`?  For utility libraries like this, the cost of sticking with good ole es5 is not worth the inability to support older versions of node.

Reading this reminded me to go back and ensure I did not use any arrow functions or `const`'s.  So I will be updating that later today hopefully.

File: test/fixtures/promises/1-callback-test.js [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+module.exports.up = function (next) {
+  setTimeout(function () {
+    return next()
+  }, 200)
+}
+
+module.exports.down = function (next) {
+  setTimeout(function () {
+    return next()
+  }, 200)
```
Comment (wesleytodd) [inline test/fixtures/promises/1-callback-test.js]: To reduce test time, just do a timeout of `1` ms.  It achieves the same effect without the wait.

File: test/promises.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+      }
+
+      assert(err)
+      assert.equal(err.message, 'Migration was already run. Please provide callback OR promise, not both.')
+      errorThrown = true
+      done()
+    })
+  })
+
+  afterEach(function (done) {
```
Comment (wesleytodd) [inline test/promises.js]: Move this to the top like in the other tests.  If this file grows it will get lost way down here.

File: lib/migrate.js [hunk 5]
```diff
@@ -26,9 +26,14 @@ function migrate (set, direction, migrationName, fn) {
     }
 
     set.emit('migration', migration, direction)
-    migration[direction](function (err) {
-      if (err) return fn(err)
+    let migrationCompleted = false
 
+    let migrationPromise = migration[direction](function (err) {
+      if (migrationCompleted) {
```
Comment (wesleytodd) [inline lib/migrate.js]: Even though I prefer multi-line if statements with curlys, the majority of the code base is done on a single line.  Please follow the common formatting.

**Comment by wesleytodd** [comment]
@jdpnielsen Awesome work.  Thanks!  With these few changes I can merge this up before we do the `1.0` release for sure.

**Comment by jdpnielsen** [comment]
Done.

**Comment by wesleytodd** [comment]
Weird, I amended the commit and added the `closes #82` but it didn't close it.  Anyway, this is published in the next beta `1.0.0-2`.  Install with `npm i migrate@next` or `npm i migrate@1.0.0-2`.

**Pull request closed without merging**